examples/serverless/main.go

```diff
@@ -0,0 +1,247 @@
+package main
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/blang/semver"
+	p "github.com/iwahbe/pulumi-go-provider"
+	r "github.com/iwahbe/pulumi-go-provider/resource"
+	apigateway "github.com/pulumi/pulumi-aws-apigateway/sdk/go/apigateway"
+	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/iam"
+	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/lambda"
+	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/s3"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+func main() {
+	err := p.Run("serverless", semver.Version{Minor: 1},
+		p.Components(
+			&Service{},
+		),
+		p.Types(
+			&Function{},
+			&ServiceProvider{},
+			&Event{},
+			&ProviderIam{},
+			&ProviderIamRole{},
+			&HttpEvent{},
+			&SqsEvent{},
+		),
+		p.PartialSpec(schema.PackageSpec{}),
+	)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s", err.Error())
+		os.Exit(1)
+	}
+}
+
+type Service struct {
+	pulumi.ResourceState
+
+	// Outputs
+	Url pulumi.StringPtrOutput `pulumi:"url" provider:"output"`
+	// Lambdas []*lambda.Function     `pulumi:"lambdas" provider:"output"`
+
+	// Inputs
+	Functions FunctionMapInput     `pulumi:"functions"`
+	Provider  ServiceProviderInput `pulumi:"provider"`
+}
+
+func (s *Service) Annotate(a r.Annotator) {
+	a.Describe(&s.Functions, "Configure the functions to deploy.")
+	a.Describe(&s.Provider, "Configure general settings to apply across all functions.")
+
+	a.Describe(&s.Url, "The URL at which any HTTP handlers are exposed, if any of the functions expose HTTP handlers.")
+}
+
+func (s *Service) Construct(name string, ctx *pulumi.Context) error {
+	component := &Service{}
+	// What about opts?
+	err := ctx.RegisterComponentResource("serverless:index:Service", name, component)
+	if err != nil {
+		return err
+	}
+
```

Comment:
```suggestion
```
`s` is the component that you are registering here. You don't need to create a new component or register it. We don't expose Options, since they should all be inherited by parenting `s`. Let me know if you can think of a reason we need to expose `Options`.